File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int max = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {

        getDiameter(root);
        
        return max;
    }
    
    public int getDiameter(TreeNode root) {
                
        if(root == null)        
            return -1;
        
        int leftHeight = getDiameter(root.left);
        int rightHeight = getDiameter(root.right);
        
        max = Math.max(leftHeight + rightHeight + 2, max);

        return Math.max(leftHeight, rightHeight) + 1;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:  

        self.height = 0      

        def dfs(root):

            if root is None:
                return 0

            self.height = max(self.height, dfs(root.left) + dfs(root.right))
            
            return max(dfs(root.left), dfs(root.right)) + 1  

        dfs(root)

        return self.height